sqlmesh/utils/__init__.py

```diff
@@ -308,11 +308,37 @@ def groupby(
     return grouped
 
 
+def columns_to_types_to_struct(
+    columns_to_types: t.Union[t.Dict[str, exp.DataType], t.Dict[str, str]]
+) -> exp.DataType:
+    """
+    Converts a dict of column names to types to a struct.
+    """
+    return exp.DataType(
+        this=exp.DataType.Type.STRUCT,
+        expressions=[
+            exp.ColumnDef(this=exp.to_identifier(k), kind=v) for k, v in columns_to_types.items()
+        ],
+        nested=True,
+    )
+
+
 def columns_to_types_all_known(columns_to_types: t.Dict[str, exp.DataType]) -> bool:
     """
     Checks that all column types are known and not NULL.
     """
-    return all(
-        not column_type.is_type(exp.DataType.Type.UNKNOWN, exp.DataType.Type.NULL)
-        for column_type in columns_to_types.values()
-    )
+
+    def check_struct(struct: exp.DataType) -> bool:
+        for column_def in struct.expressions:
+            if column_def.kind.is_type(exp.DataType.Type.UNKNOWN, exp.DataType.Type.NULL):
+                return False
+            if column_def.kind.is_type(exp.DataType.Type.STRUCT):
+                if not check_struct(column_def.kind):
+                    return False
+            if column_def.kind.is_type(exp.DataType.Type.ARRAY):
+                if column_def.kind.expressions[0].is_type(exp.DataType.Type.STRUCT):
```

Comment:
expressions[0] may not exist? this may be an index error

Reply:
Does it not have to exist if it is an array?

Reply:
i think it's possible to just have an ArrayType, not sure if it's practical, but the code path exists

Reply:
Ok I can just check and return false if not

Reply:
Ended up rewriting a bit to orient around expressions instead of structs since we don't care about column names (like in SchemaDiffer).